Watchdog based reloader

tilgovi opened the pull request:
Replace the reloader with a reloader based on the watchdog package.
This new reloader should work on all the platforms. Most popular
platforms use an underlying event system like inotify or kqueue, but
watchdog will fall back to a polling implementation.

Start the reloader after loading the WSGI entry point of the application
so that the reloader picks up the application code modules when scanning
sys.modules and have the worker send SIGQUIT to itself in its callback
because that seems to make exits happen more quickly, which is more in
line with the common development use case for reloading.

tilgovi commented:
There are probably some bits I missed, documentation and whatnot, but I wanted to get this out here for feedback.

tilgovi commented:
Closes #1477
Closes #1537

Might also fix other problems that have been reported with the reloader, like #1494 and #1562.

tilgovi commented:
Indeed :-). Fixed.

sevein commented:
@tilgovi, will this work solve #1494?

tilgovi commented:
I don't know, but I would love to find out! Please comment if you test it.

berkerpeksag commented on gunicorn/_reloader.py:
```diff
@@ -0,0 +1,58 @@
+        self._handler = ReloadingEventHandler(callback)
+        self._dirs = set()
+
+        for filename in (extra_files or []):
```
Just a nitpick, but adding an `if extra_files is not None:` branch might be more readable.

berkerpeksag commented on gunicorn/config.py:
```diff
@@ -868,9 +864,13 @@ class ReloadEngine(Setting):
+
+        .. versionchanged:: 19.8
+           A value of 'inotify' is an alias for 'auto' and uses the
+           best implementation available on the platform.
```
Should we also explicitly deprecate `'inotify'` value and remove it in 20?

tilgovi replied:
We could, but I'm not sure it's worth the disruption, however minor. It's harmless to me to have it as an alias.

tilgovi replied:
You think I should remove the option to pick and always try to use the best? I'm okay with that, too.

tilgovi replied:
But I also would like to ship this before R20.

berkerpeksag commented on gunicorn/reloader.py:
```diff
@@ -3,128 +3,15 @@
+except ImportError:
+    class Reloader(object):
+        def __init__(self, *args, **kwargs):
+            raise ImportError('You must have the watchdog module '
```
`gunicorn.workers.gaiohttp` raises `RuntimeError`: 

https://github.com/benoitc/gunicorn/blob/0e4e0779e63c080777bfddb02de4cb828cfdb1c7/gunicorn/workers/gaiohttp.py#L12

Should we follow it or raise `ImportError` instead?

kennethreitz replied:
Or, can we just make watchdog a dependency?

kennethreitz replied:
I'd like to encourage local development with gunicorn, if possible.

tilgovi replied:
I'm not opposed but I'll defer to @benoitc.

tilgovi replied:
I will make it a RuntimeError if we can't quickly vendor watchdog.

kennethreitz commented:
This sounds nice :)

tilgovi commented on gunicorn/_reloader.py:
```diff
@@ -0,0 +1,58 @@
+        if engine == 'auto' or engine == 'inotify':
+            self._observer = Observer()
+        else:
+            assert engine == 'poll'
```
I left it as an assert because it should be caught be the configuration system, but I can change it if it still doesn't seem right to you.

kennethreitz replied:
Indeed, a few people do use `-O` in production.

benoitc replied:
@tilgovi "it should be caught by the configuration system" what do you mean?

benoitc replied:
I looked at the issue list of the project and it seems it doesn't work everywhere. Also it doesn't sound very active. Before taking any decision i would check i it really works multi-platform as some issues let me think it is not (like on docker or osx). 

Anyway i'm not inclined to make such thing an hard dependency. Though it would be awesome to handle local reloads and not only for local developments. Silly idea but couldn't have equivalent or fork maintained directly inside gunicorn?

tilgovi replied:
It looks like https://github.com/gorakhargosh/watchdog is the repo.

It does seem that the last commits are last November, but the issues that people seem to open recently don't seem too urgent. Although, I see some good looking PRs that are stale. I would rather try to get in touch with the maintainer(s) and see if we can unblock things than to fork it into Gunicorn.

It does claim to support all these platforms, and I believe it. At the very least, it supports a fallback that works like ours did. On every platform except BSD, it seems, there's fallbacks to the polling observer: https://github.com/gorakhargosh/watchdog/blob/master/src/watchdog/observers/__init__.py

tilgovi replied:
Is your objection to a hard dependency about code size or security or bugs or... we can code defensively and make the PollingObserver the default if that feels safer.

tilgovi replied:
I've e-mailed the most recent maintainers to ask the status of the project.

benoitc replied:
@tilgovi did you have a response?

For me automatic reload support should be a core feature of Gunicorn (like HUP is). We should even improve it by offering a good support of gunicorn for the editors and IDEs but also offere automatic reload support for prod like the one you find in caddy.

 **If we consider it is a core feature and useful for our user, then we should be able to support it.** 

I'm quite doubtful we could offer a good support of it by using a dependency. We already have some issues with some of the dependency inside gunicorn even indirect that need to be fixed. 

That's why I proposed to maybe fork it and use our own package/ internal source code (could we use submodules and still patch upstream?). In between we could a good default (internal obsever) and allow the use of watchdog if available (and maybe encourage with a log) like you propose.  Since we are going to full python 3 soon, support of kqueue and other should be easier.

Thoughts?

tilgovi replied:
I think you're right. I have not heard back from my e-mail to the watchdog maintainers. We do need this to work. We should vendor it. Maybe @kennethreitz has good thoughts because I know there is vendoring in pipenv. Rather than just copy modules into the gunicorn package, we should copy a whole package as much as possible and fork. Maybe this is a good use of the gunicorn github org?

tilgovi commented:
I tried to address all the comments. Please weigh in on making watchdog required or anything else. Thanks!

aconrad commented:
Hey @tilgovi, would you mind rebasing since coverage reports should be fixed in master?

Code0x58 commented on gunicorn/_reloader.py:
```diff
@@ -0,0 +1,59 @@
+        dirname = os.path.dirname(filename)
+        self._observer.schedule(self._handler, dirname)
+
+    def get_dirs(self):
```
I don't like gunicorn's approach attempt at being smart about reloading, I think it would be better to take directories to watch recursively (and add directories which are created), as the current watching leaves a lot of [holes](https://github.com/benoitc/gunicorn/issues/1562#issuecomment-326415960). Then the user can have their project directory watched for changes (including resources like templates and static content).

tilgovi commented on gunicorn/workers/base.py:
```diff
@@ -111,21 +111,20 @@ def init_process(self):
-                self.cfg.worker_int(self)
-                time.sleep(0.1)
-                sys.exit(0)
+                os.kill(self.pid, signal.SIGQUIT)
```
No, SIGQUIT causes the same code to run. I thought it was cleaner to send the signal.

tilgovi commented on gunicorn/_reloader.py:
```diff
@@ -0,0 +1,59 @@
+
+    def get_dirs(self):
+        fnames = [
+            os.path.dirname(re.sub('py[co]$', 'py', module.__file__))
```
Can you explain this more? I did take the same code as the inotify reloader, but I don't know what mistake it has.

Code0x58 replied:
`os.path.dirname` extracts the directory's path from a given path, but the regex substitution modifies the file name (`os.path.basename`), so it has no effect on the return value.

tilgovi replied:
Thanks! I'll remove that.

tilgovi commented:
Updated the branch:

- rebased onto latest master
- changed `ImportError` to `RuntimeError`
- removed all changes other than to the reloader instantiation

I think it's cleaner to have the worker send itself SIGQUIT because then exit is handled from the main thread rather than a reloader thread, but I'll make that a separate PR.

I think the reloader should be started after loading the wsgi callable to ensure as much application code as possible is loaded, but I'll make that a separate PR.

For a 19.x release, I would rather leave 'inotify' as an alias for 'auto' so that there is no breaking change at all. For R20, we can make it very explicit if we want to, depending on how we vendor watchdog or some of its modules.

@Code0x58 if you could point me to an issue for the directory/filename mistake, I would appreciate that.

benoitc commented:
where is my comment on that PR? I didn't have any answer on it :(

@tilgovi did you have any answer from the maintainer of the watchdog module?

 I also still want addressed the question of support and how core the reload feature should be. Should it be discussed in a separate ticket?

philfreo commented:
+1 for getting this finished/merged. @tilgovi Are you still up for finishing it?

tilgovi commented:
@philfreo I am still up for it, but would be happy to let anyone else take it to completion, as well. Are you interested?

tilgovi commented:
I think we still have a couple open questions:

- Watchdog seems unmaintained. Do we want it as a dependency?
- Should this be core or extra functionality (either via package `extras` or a separate package)?

philfreo commented:
Please do handle it then - you'd be much better than I :) Just glad to see it moving forward.

My 2c: if the [`reload` option](http://docs.gunicorn.org/en/stable/settings.html#reload) is part of core, then it seems like this should be too?

mmerickel commented:
The pylons project recently replaced it's thread-based reloader with `hupper` [1] which does a more robust reloading using either polling or watchdog (if installed). I'd be happy to help with issues if anyone made an attempt to integrate it here.

[1] https://docs.pylonsproject.org/projects/hupper/en/latest/

tilgovi commented:
I would love to replace this with `hupper`. I think `hupper` is a better dependency than `watchdog` itself. We have managed not to introduce any out-of-tree dependencies in Gunicorn so far, but maybe @benoitc would consider a dependency on `hupper`. It could even be optional.

benoitc commented:
go ahead i missed that.

tilgovi commented:
Let's revisit after #1791 and try with `hupper`.

tilgovi commented:
Actually, not sure #1791 is still applicable. I will check.

benoitc commented:
what happened to that PR ? Is this still something we can work on @tilgovi ?

tilgovi commented:
I think I was going to try to use `hupper` instead of `watchdog`, if it's acceptable to add a dependency. Maybe make it optional? What do you think?

n1ngu commented:
My 2 cents:
- Any reloading dependency should be made optional as an extra requirement. On runtime, silently downgrade the reloader implementation if some import is missing. Maybe raise a warning?
- Watchdog is a rather ubiquitous package. I've seen it integrated with werkzeug/flask, django and other development servers. Everybody is happy with it and its a stable component from the python ecosystem so... I do not understand the reluctance to introduce it.
- AFAIU `hupper`'s purpose is to totally replace any reloading hassle from other python programs. So, without further ado: today, right now, with any past gunicorn version, without any integration in gunicorn's side, anyone could run `hupper -m gunicorn application:app` instead of `gunicorn --reload application:app`. There would be nothing to implement in our side but dropping the `--reload` option and advising people to `pip install hupper` instead. All the complexity about choosing the best reloader for a given platform depending on the presence/absence of watchdog or anything else would be just moved away without having to integrate anything. Yet, this is a much more marginal project than watchdog so I can't really vouch for this (out of unfamiliarity with hupper).
